pymeasure/instruments/keithley/keithley2600.py

```diff
@@ -0,0 +1,327 @@
+        """ Configures the measurement of current.
+        :param nplc: Number of power line cycles (NPLC) from 0.001 to 25
+        :param current: Upper limit of current in Amps, from -1.5 A to 1.5 A
+        :param auto_range: Enables auto_range if True, else uses the set current
+        """
+        log.info("%s is measuring current." % self.channel)
+        self.instrument.write('smu%s.measure.i()' % self.channel)
+        self.instrument.write('smu%s.measure.nplc=%f' % (self.channel, nplc))
+        if auto_range:
+            self.instrument.write('smu%s.measure.autorangei=1' % self.channel)
+        else:
+            self.current_range = current
+        self.check_errors()
+
+    def auto_range_source(self):
+        """ Configures the source to use an automatic range.
+        """
+        if self.source_mode == 'current':
+            self.instrument.write('smu%s.source.autorangei=1' % self.channel)
+        else:
+            self.instrument.write('smu%s.source.autorangev=1' % self.channel)
+
+    def apply_current(self, current_range=None, compliance_voltage=0.1):
+        """ Configures the instrument to apply a source current, and
+                uses an auto range unless a current range is specified.
+                The compliance voltage is also set.
+                :param compliance_voltage: A float in the correct range for a
+                                           :attr:`~.Keithley2600.compliance_voltage`
+                :param current_range: A :attr:`~.Keithley2600.current_range` value or None
+                """
+        log.info("%s is sourcing current." % self.channel)
+        self.source_mode = 'current'
+        if current_range is None:
+            self.auto_range_source()
+        else:
+            self.source_current_range = current_range
+        self.compliance_voltage = compliance_voltage
+        self.check_errors()
+
+    def apply_voltage(self, voltage_range=None,
+                      compliance_current=0.1):
+        """ Configures the instrument to apply a source voltage, and
+        uses an auto range unless a voltage range is specified.
+        The compliance current is also set.
+        :param compliance_current: A float in the correct range for a
+                                   :attr:`~.Keithley2600.compliance_current`
+        :param voltage_range: A :attr:`~.Keithley2600.voltage_range` value or None
+        """
+
+        log.info("%s is sourcing voltage." % self.channel)
+
+        self.source_mode = 'voltage'
+        if voltage_range is None:
+            self.auto_range_source()
+        else:
+            self.source_voltage_range = voltage_range
+        self.compliance_current = compliance_current
+        self.check_errors()
+
+    @property
+    def error(self):
+        """ Returns a tuple of an error code and message from a
+        single error. """
+
+        err = self.instrument.ask('print(errorqueue.next())')
```

Comment:
what is returned if there is no error, or the queue is empty? can your parsing below deal with that correctly?

Reply:
Am I interpreting correctly that the error queue is not per-channel, but for the whole instrument? 
If so, similar to the reasoning above, it would be preferable to 
* move the `error` and `check_errors` implementations to the instrument, and 
* in the Channel define a method check_errors (some properties expect that to be present), that just calls the one in the instrument and returns the result.